### PR TITLE
Tank factory

### PR DESCRIPTION
**Pull request by Glebeeezy**
Homework done.

### CODE REVIEW

File: TankFactory/Contracts/TankBase.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            if (_engineIsRunning == false)
+            {
+                _engineIsRunning = true;
+            }
+        }
+
+        public void StopEngine()
+        {
+            if (_engineIsRunning == true)
```
Comment (HolubVasy) [inline TankFactory/Contracts/TankBase.cs:51]: _!Comment_

### Code concise

`== true` can be safely removed.

File: TankFactory/Realisations/HeavyTank.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        private new int _speed = 20;
+
+        public override void Fire()
+        {
+            Console.WriteLine("Бабах");
+        }
+
+        public override void StartMoving(int timeOfMoivng)
+        {
+            if (_engineIsRunning == true)
```
Comment (HolubVasy) [inline TankFactory/Realisations/HeavyTank.cs:17]: _!Comment_

### Code concise

`== true` can be safely removed.

File: TankFactory/Realisations/HeavyTank.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                        Y = Y + _speed;
+                    }
+                    StopMoving();
+                }
+                else if (_direction == Direction.right && _isMoving == true)
+                {
+                    for (int i = 0; i < timeOfMoivng; i++)
+                    {
+                        X = X + _speed;
+                    }
```
Comment (HolubVasy) [inline TankFactory/Realisations/HeavyTank.cs:34]: _!Comment_

### Code concise

`== true` can be safely removed.

File: TankFactory/Realisations/LightTank.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            if (_engineIsRunning == true)
+            {
+                _isMoving = true;
+
+                if (_direction == Direction.straight && _isMoving == true)
+                {
+                    for (int i = 0; i < timeOfMoivng; i++)
+                    {
+                        Y = Y + _speed;
```
Comment (HolubVasy) [inline TankFactory/Realisations/LightTank.cs:18]: _!Recommendation_

### DRY

Consider creation of private method for repeated logic and reuse it.

```
for (int i = 0; i < timeOfMoivng; i++)
{
    X = X + _speed;
}
StopMoving();
```

File: TankFactory/Realisations/MediumTank.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                _isMoving = true;
+
+                if (_direction == Direction.straight && _isMoving == true)
+                {
+                    for (int i = 0; i < timeOfMoivng; i++)
+                    {
+                        Y = Y + _speed;
+                    }
+                    StopMoving();
```
Comment (HolubVasy) [inline TankFactory/Realisations/MediumTank.cs:27]: _!Recommendation_

### DRY

Consider creation of private method for repeated logic and reuse it.

```
for (int i = 0; i < timeOfMoivng; i++)
{
    X = X + _speed;
}
StopMoving();
```

part.

**Review by HolubVasy — CHANGES_REQUESTED** [review]

File: TankFactory/Contracts/TankBase.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        private string _tankType;
+
+        protected string TankType
+        {
+            get { return _tankType; }
+            init { _tankType = value; }
+        }
+
+
+        protected bool _engineIsRunning = false;
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Contracts/TankBase.cs:19]: Default value for bool is false

File: TankFactory/Contracts/TankBase.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+        protected bool _isMoving = false;
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Contracts/TankBase.cs:21]: Default value for bool is false. '= false' can be safely removed

**Review by VasiliHolubITAcademy2023 — APPROVED** [review]

File: TankFactory/Realisations/HeavyTank.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            for (int i = 0; i < timeOfMoving; i++)
+            {
+                X = X + _speed;
+            }
+            StopMoving();
+        }
+
+        private void MoveLeft(int timeOfMoving)
+        {
+            for (int i = 0; i < timeOfMoving; i++)
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Realisations/HeavyTank.cs:64]: Can be converted into `Enumerable.Range(0,  timeOfMoving  )` for using foreach construction

File: TankFactory/Realisations/LightTank.cs [hunk 9]
```diff
@@ -0,0 +1,78 @@
+﻿using TankFactory.Contracts;
+using TankFactory.Enums;
+
+namespace TankFactory
+{
+    internal partial class LightTank : TankBase
+    {
+        public override void StartMoving(int timeOfMoivng)
+        {
+            if (_engineIsRunning)
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Realisations/LightTank.cs:10]: Try using switch construction

File: TankFactory/StaticFabric/TankFabric.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+                return tank;
+            }
+
+            else if(!string.IsNullOrEmpty(tankType) && tankType == "HeavyTank")
+            {
+                HeavyTank tank = new HeavyTank(tankType);
+                return tank;
+            }
+
+            else
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/StaticFabric/TankFabric.cs:33]: Extract validation and code guard in the start of method for conciseness 

```
if ( string.IsNullOrEmpty ( typeType )  )
{
      Console.WriteLine("Указан неверный тип танка");
      return null;
}

```

File: TankFactory/Contracts/TankBase.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+        protected int _speed = 10;
+
+        private int _x;
+
+        public int X
+        {
+            get { return _x; }
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Contracts/TankBase.cs:29]: Can be safely converted in to auto-property as `_x` is not used directly

File: TankFactory/Contracts/TankBase.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            protected set { _x = value; }
+        }
+
+        private int _y;
+
+        public int Y
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Contracts/TankBase.cs:35]: Can be safely converted in to auto-property as `_y` is not used directly

File: TankFactory/Contracts/TankBase.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        protected void StopMoving()
+        {
+            _isMoving = false;
+        }
+
+        public void TurnLeft()
+        {
+            if (_engineIsRunning)
+            {
+                if (_direction == Direction.straight)
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Contracts/TankBase.cs:68]: Try use switch by occasion

File: TankFactory/Enums/DirectionsEnum.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (VasiliHolubITAcademy2023) [inline TankFactory/Enums/DirectionsEnum.cs:3]: Remove not used using expressions

**Pull request merged**